RecoTracker/LSTCore/src/alpaka/MiniDoublet.h

```diff
@@ -0,0 +1,1093 @@
+                                                   noShiftedDz,
+                                                   noShiftedDphi,
+                                                   noShiftedDphiChange,
+                                                   xLower,
+                                                   yLower,
+                                                   zLower,
+                                                   rtLower,
+                                                   xUpper,
+                                                   yUpper,
+                                                   zUpper,
+                                                   rtUpper);
+          if (success) {
+            int totOccupancyMDs =
+                alpaka::atomicOp<alpaka::AtomicAdd>(acc, &mdsInGPU.totOccupancyMDs[lowerModuleIndex], 1u);
+            if (totOccupancyMDs >= (rangesInGPU.miniDoubletModuleOccupancy[lowerModuleIndex])) {
+#ifdef Warnings
+              printf("Mini-doublet excess alert! Module index =  %d\n", lowerModuleIndex);
+#endif
+            } else {
+              int mdModuleIndex = alpaka::atomicOp<alpaka::AtomicAdd>(acc, &mdsInGPU.nMDs[lowerModuleIndex], 1u);
+              unsigned int mdIndex = rangesInGPU.miniDoubletModuleIndices[lowerModuleIndex] + mdModuleIndex;
+
+              addMDToMemory(acc,
+                            mdsInGPU,
+                            hitsInGPU,
+                            modulesInGPU,
+                            lowerHitArrayIndex,
+                            upperHitArrayIndex,
+                            lowerModuleIndex,
+                            dz,
+                            dphi,
+                            dphichange,
+                            shiftedX,
+                            shiftedY,
+                            shiftedZ,
+                            noShiftedDz,
+                            noShiftedDphi,
+                            noShiftedDphiChange,
+                            mdIndex);
+            }
+          }
+        }
+      }
+    }
+  };
+
+  struct createMDArrayRangesGPU {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                  struct SDL::modules modulesInGPU,
+                                  struct SDL::objectRanges rangesInGPU) const {
+      auto const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+      auto const gridThreadExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+
+      // Initialize variables in shared memory and set to 0
+      int& nTotalMDs = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      nTotalMDs = 0;
+      alpaka::syncBlockThreads(acc);
+
+      // Initialize variables outside of the for loop.
+      int occupancy, category_number, eta_number;
+
+      for (uint16_t i = globalThreadIdx[2]; i < *modulesInGPU.nLowerModules; i += gridThreadExtent[2]) {
```

Comment:
what's a good way/syntax to implement a kernel that will work only with `Acc1D`?
I added a `static_assert(is_same_v(TAcc, Acc1D), ` at the top.

Reply:
If you rely on the (backend-specific) `Acc1D` type alias, you could just make the function a non-template
```cpp
    ALPAKA_FN_ACC void operator()(Acc1D const& acc,
                                  struct lst::Modules modulesInGPU,
                                  struct lst::ObjectRanges rangesInGPU) const {
```
IIRC this pattern is not (yet) used elsewhere in CMSSW, but is discussed in https://github.com/cms-sw/cmssw/issues/44625. It would require moving the code inside `ALPAKA_ACCELERATOR_NAMESPACE` (but you should do that anyway if you use any of the type aliases defined there).

Alternatively, I _believe_ (but didn't test) a generic test on the dimensionality of the `TAcc` could be done along
```cpp
static_assert(alpaka::Dim<TAcc>::value == 1u);
// or
static_assert(std::is_same_v<alpaka::Dim<TAcc>, alpaka_common::Dim1D>);
```
(not sure if such checks could be implemented as part of SFINAE, or C++20 concepts now that we can, at least theoretically, use them)

Just to think out loud, elsewhere our habit has been to not check the dimensionality of the `Acc` in the kernel, perhaps we should make such constraints more explicit.

Reply:
> If you rely on the (backend-specific) `Acc1D` type alias, you could just make the function a non-template

I do like this approach.



> Alternatively, I _believe_ (but didn't test) a generic test on the dimensionality of the `TAcc` could be done

Yes, that works.



> Just to think out loud, elsewhere our habit has been to not check the dimensionality of the `Acc` in the kernel, perhaps we should make such constraints more explicit.

Noted.